kafka/signed_blinded_token_issuer_handler.go

```diff
@@ -24,135 +26,209 @@ func SignedBlindedTokenIssuerHandler(
+			errors.New(message),
+			message,
+			PERMANENT,
+		)
+	}
+
+	if request.Issuer_cohort != 0 && request.Issuer_cohort != 1 {
+		message := fmt.Sprintf(
+			"Request %s: Provided cohort is not supported: %d",
+			requestId,
+			request.Issuer_cohort,
+		)
+		errorChannel <- batgo_handlers.WrapError(
+			errors.New(message),
+			message,
+			PERMANENT,
+		)
+	}
+
+	issuer, appErr := server.GetLatestIssuer(request.Issuer_type, int(request.Issuer_cohort))
+	if appErr != nil {
+		blindedTokenResults <- avroSchema.SigningResult{
+			Signed_tokens:     nil,
+			Issuer_public_key: "",
+			Status:            INVALID_ISSUER,
+			Associated_data:   request.Associated_data,
 		}
-		// @TODO: If one token fails they will all fail. Assess this behavior
-		signedTokens, dleqProof, err := btd.ApproveTokens(blindedTokens, issuer.SigningKey)
+	}
+
+	var blindedTokens []*crypto.BlindedToken
+	// Iterate over the provided tokens and create data structure from them,
+	// grouping into a slice for approval
+	for _, stringBlindedToken := range request.Blinded_tokens {
+		blindedToken := crypto.BlindedToken{}
+		err := blindedToken.UnmarshalText([]byte(stringBlindedToken))
 		if err != nil {
-			logger.Error().Msg(fmt.Sprintf("Request %s: Could not approve new tokens: %e", blindedTokenRequestSet.Request_id, err))
-			blindedTokenResults = append(blindedTokenResults, avroSchema.SigningResult{
+			logger.Error().Msg(fmt.Sprintf("Request %s: failed to unmarshal blinded tokens: %e", requestId, err))
+			blindedTokenResults <- avroSchema.SigningResult{
 				Signed_tokens:     nil,
 				Issuer_public_key: "",
 				Status:            ERROR,
 				Associated_data:   request.Associated_data,
-			})
-			continue
-		}
-		marshaledDLEQProof, err := dleqProof.MarshalText()
-		if err != nil {
-			return errors.New(
-				fmt.Sprintf(
-					"Request %s: Could not marshal DLEQ proof: %e",
-					blindedTokenRequestSet.Request_id,
-					err,
-				),
-			)
-		}
-		var marshaledTokens []string
-		for _, token := range signedTokens {
-			marshaledToken, err := token.MarshalText()
-			if err != nil {
-				return errors.New(
-					fmt.Sprintf(
-						"Request %s: Could not marshal new tokens to bytes: %e",
-						blindedTokenRequestSet.Request_id,
-						err,
-					),
-				)
 			}
-			marshaledTokens = append(marshaledTokens, string(marshaledToken[:]))
 		}
-		publicKey := issuer.SigningKey.PublicKey()
-		marshaledPublicKey, err := publicKey.MarshalText()
+		blindedTokens = append(blindedTokens, &blindedToken)
+	}
+	// @TODO: If one token fails they will all fail. Assess this behavior
+	signedTokens, dleqProof, err := btd.ApproveTokens(blindedTokens, issuer.SigningKey)
+	if err != nil {
+		logger.Error().Msg(fmt.Sprintf("Request %s: Could not approve new tokens: %e", requestId, err))
+		blindedTokenResults <- avroSchema.SigningResult{
+			Signed_tokens:     nil,
+			Issuer_public_key: "",
+			Status:            ERROR,
+			Associated_data:   request.Associated_data,
+		}
```

Comment:
Final question, final question! I won’t bother more 😆

Why sending a result with an error and not an error to the error channel?

Given that all of the tokens will fail.


(I am missing so much context… This is just out of curiosity)